File: v4/rust/iprange-cli/src/rpc/handlers/convert.rs
//! Mechanical public-SDK-value conversion for JSON-RPC results.

use iprange_livedb::{
    AddressFamily, DatabaseInfo, MetaSelection, NetworkEnrichmentV1, NetworkEnrichmentV1Location,
    NetworkEnrichmentV1View, StructureKind, ValueKind,
};
use serde_json::{json, Value};

use super::super::state::CursorPoint;

pub fn decimal_u64(value: u64) -> String {
    value.to_string()
}

pub(crate) fn hex_bytes(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

pub fn hex_id(bytes: &[u8; 16]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

pub fn value_tag(bytes: &[u8]) -> Value {
    json!({ "hex": bytes.iter().map(|byte| format!("{byte:02x}")).collect::<String>() })
}

pub fn address_family(value: AddressFamily) -> &'static str {
    match value {
        AddressFamily::Ipv4 => "ipv4",
        AddressFamily::Ipv6 => "ipv6",
    }
}

pub fn value_kind(value: ValueKind) -> &'static str {
    match value {
        ValueKind::Direct => "direct",
        ValueKind::Membership => "membership",
        ValueKind::Structured => "structured",
    }
}

pub fn structure_kind(value: StructureKind) -> &'static str {
    match value {
        StructureKind::None => "none",
        StructureKind::NetworkEnrichmentV1 => "network_enrichment_v1",
    }
}

pub fn meta_selection(value: MetaSelection) -> &'static str {
    match value {
        MetaSelection::ProvenCurrent => "proven_current",
        MetaSelection::SoleMeta0 => "sole_meta_0",
        MetaSelection::SoleMeta1 => "sole_meta_1",
    }
}

pub fn database_info(info: &DatabaseInfo) -> Value {
    json!({
        "address_family": address_family(info.address_family),
        "value_kind": value_kind(info.value_kind),
        "structure_kind": structure_kind(info.structure_kind),
        "value_tag": value_tag(info.value_tag.bytes()),
        "database_id": hex_id(&info.database_id),
        "transaction_id": decimal_u64(info.transaction_id),
        "commit_nonce": hex_id(&info.commit_nonce),
        "page_count": decimal_u64(info.page_count),
        "range_record_count": decimal_u64(info.range_record_count),
        "active_feed_count": decimal_u64(info.active_feed_count),
        "meta_selection": meta_selection(info.meta_selection),
    })
}

pub fn location(value: NetworkEnrichmentV1Location) -> Value {
    json!({
        "latitude_microdegrees": value.latitude_microdegrees,
        "longitude_microdegrees": value.longitude_microdegrees,
    })
}

pub fn network_enrichment(value: NetworkEnrichmentV1, threat_feeds: &[String]) -> Value {
    let location = value.location.map(location).unwrap_or_else(|| Value::Null);
    json!({
        "asn": value.asn,
        "country_id": value.country_id,
        "state_id": value.state_id,
        "city_id": value.city_id,
        "location": location,
        "threat_feeds": threat_feeds,
    })
}

pub fn cursor_address(point: CursorPoint) -> String {
    match point {
        CursorPoint::V4(value) => format!(
            "{}.{}.{}.{}",
            value >> 24,
            (value >> 16) & 0xff,
            (value >> 8) & 0xff,
            value & 0xff
        ),
        CursorPoint::V6(value) => std::net::Ipv6Addr::from(value.to_be_bytes()).to_string(),
    }
}

/// Convert a kind-1/kind-2 SDK local identity to its documented volume/file
/// decimal pair. Unsupported identity kinds remain handler errors.
#[cfg(test)]
pub fn file_identity(
    identity: &iprange_livedb::validation::LocalFileIdentity,
) -> Result<Value, ()> {
    let tail = &identity.bytes[16..];
    let device = u64::from_le_bytes(identity.bytes[0..8].try_into().map_err(|_| ())?);
    let file = u64::from_le_bytes(identity.bytes[8..16].try_into().map_err(|_| ())?);
    match identity.kind {
        1 if tail.iter().all(|byte| *byte == 0) => Ok(json!({
            "volume": decimal_u64(device),
            "file": decimal_u64(file),
        })),
        2 if identity.bytes[24..].iter().all(|byte| *byte == 0) => Ok(json!({
            "volume": decimal_u64(device),
            "file": decimal_u64(file),
        })),
        _ => Err(()),
    }
}

pub fn enrichment_view<'a>(view: &NetworkEnrichmentV1View<'a>, threat_feeds: &[String]) -> Value {
    network_enrichment(view.value(), threat_feeds)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn exact_and_identity_conversions() {
        assert_eq!(decimal_u64(u64::MAX), "18446744073709551615");
        let mut identity = [0u8; 16];
        identity[..4].copy_from_slice(&[0x0a, 0xbc, 0xde, 0xf0]);
        assert_eq!(hex_id(&identity), format!("0abcdef0{}", "0".repeat(24)));
        assert_eq!(value_tag(b"direct"), json!({"hex": "646972656374"}));
    }

    #[test]
    fn kind_one_file_identity_decodes_little_endian_pair() {
        let mut bytes = [0u8; 32];
        bytes[..8].copy_from_slice(&0x0102030405060708u64.to_le_bytes());
        bytes[8..16].copy_from_slice(&0x090a0b0c0d0e0f10u64.to_le_bytes());
        let identity = iprange_livedb::validation::LocalFileIdentity { kind: 1, bytes };
        assert_eq!(
            file_identity(&identity).unwrap(),
            json!({"volume": "72623859790382856", "file": "651345242494996240"})
        );
    }

    #[test]
    fn unsupported_file_identity_kind_is_rejected() {
        let identity = iprange_livedb::validation::LocalFileIdentity {
            kind: 9,
            bytes: [1; 32],
        };
        assert_eq!(file_identity(&identity), Err(()));
    }

    #[test]
    fn cursor_addresses_are_canonical() {
        assert_eq!(cursor_address(CursorPoint::V4(0xc0000201)), "192.0.2.1");
        assert_eq!(
            cursor_address(CursorPoint::V6(0x20010db8000000000000000000000001)),
            "2001:db8::1"
        );
    }
}

File: v4/go/internal/cli/handlers/convert.go
// Mechanical public-SDK-value conversion for JSON-RPC results (Rust
// handlers/convert.rs parity).

package handlers

import (
	"fmt"
	"net/netip"

	iprangedb "github.com/firehol/iprange/v4/go"
	"github.com/firehol/iprange/v4/go/internal/cli/rpc"
)

// DecimalUint renders a u64 as a decimal wire string.
func DecimalUint(value uint64) string {
	return fmt.Sprintf("%d", value)
}

// HexBytes renders bytes as lowercase hex (wire digest/id form).
func HexBytes(bytes []byte) string {
	const digits = "0123456789abcdef"
	out := make([]byte, 0, len(bytes)*2)
	for _, b := range bytes {
		out = append(out, digits[b>>4], digits[b&0x0f])
	}
	return string(out)
}

// HexID renders a 16-byte identity as lowercase hex.
func HexID(bytes *[16]byte) string {
	return HexBytes(bytes[:])
}

// ValueTagJSON renders a value tag as its wire object.
func ValueTagJSON(tag iprangedb.ValueTag) map[string]any {
	return map[string]any{"hex": HexBytes(tag.Bytes())}
}

// AddressFamilyName maps the SDK family to its wire name.
func AddressFamilyName(family iprangedb.AddressFamily) string {
	switch family {
	case iprangedb.AddressFamilyIPv4:
		return "ipv4"
	case iprangedb.AddressFamilyIPv6:
		return "ipv6"
	}
	return "unknown"
}

// ValueKindName maps the SDK value kind to its wire name.
func ValueKindName(kind iprangedb.ValueKind) string {
	switch kind {
	case iprangedb.ValueKindDirect:
		return "direct"
	case iprangedb.ValueKindMembership:
		return "membership"
	case iprangedb.ValueKindStructured:
		return "structured"
	}
	return "unknown"
}

// StructureKindName maps the SDK structure kind to its wire name.
func StructureKindName(kind iprangedb.StructureKind) string {
	switch kind {
	case iprangedb.StructureKindNone:
		return "none"
	case iprangedb.StructureKindNetworkEnrichmentV1:
		return "network_enrichment_v1"
	}
	return "unknown"
}

// MetaSelectionName maps the SDK meta selection to its wire name.
func MetaSelectionName(selection iprangedb.MetaSelection) string {
	switch selection {
	case iprangedb.MetaSelectionProvenCurrent:
		return "proven_current"
	case iprangedb.MetaSelectionSoleMeta0:
		return "sole_meta_0"
	case iprangedb.MetaSelectionSoleMeta1:
		return "sole_meta_1"
	}
	return "unknown"
}

// DatabaseInfoJSON converts the SDK database info to the wire result
// object.
func DatabaseInfoJSON(info *iprangedb.DatabaseInfo) map[string]any {
	return map[string]any{
		"address_family":     AddressFamilyName(info.Family),
		"value_kind":         ValueKindName(info.ValueKind),
		"structure_kind":     StructureKindName(info.StructureKind),
		"value_tag":          ValueTagJSON(info.ValueTag),
		"database_id":        HexID(&info.DatabaseID),
		"transaction_id":     DecimalUint(info.TransactionID),
		"commit_nonce":       HexID(&info.CommitNonce),
		"page_count":         DecimalUint(info.PageCount),
		"range_record_count": DecimalUint(info.RangeRecordCount),
		"active_feed_count":  DecimalUint(info.ActiveFeedCount),
		"meta_selection":     MetaSelectionName(info.MetaSelection),
	}
}

// LocationJSON converts one optional network-enrichment location to
// its wire value (null when absent).
func LocationJSON(location iprangedb.NetworkEnrichmentV1Location, has bool) any {
	if !has {
		return nil
	}
	return map[string]any{
		"latitude_microdegrees":  location.LatitudeMicrodegrees,
		"longitude_microdegrees": location.LongitudeMicrodegrees,
	}
}

// NetworkEnrichmentJSON converts one decoded structured payload to its
// wire object.
func NetworkEnrichmentJSON(value iprangedb.NetworkEnrichmentV1, threatFeeds []string) map[string]any {
	return map[string]any{
		"asn":          value.ASN,
		"country_id":   value.CountryID,
		"state_id":     value.StateID,
		"city_id":      value.CityID,
		"location":     LocationJSON(value.Location, value.HasLocation),
		"threat_feeds": threatFeeds,
	}
}

// CursorAddress renders one canonical cursor checkpoint address.
func CursorAddress(point rpc.CursorPoint) string {
	if point.V4 != nil {
		return fmt.Sprintf("%d.%d.%d.%d", *point.V4>>24, (*point.V4>>16)&0xff, (*point.V4>>8)&0xff, *point.V4&0xff)
	}
	if point.V6 != nil {
		return ipv6String(*point.V6)
	}
	return ""
}

// CursorAddressV6 renders an IPv6 checkpoint address (a u128 in the
// wire format) canonically.
func CursorAddressV6(v6 *iprangedb.IPv6) string {
	if v6 == nil {
		return ""
	}
	return ipv6String(*v6)
}

// ipv6String renders the Hi/Lo halves as canonical IPv6 text.
func ipv6String(v6 iprangedb.IPv6) string {
	var bytes [16]byte
	bytes[0] = byte(v6.Hi >> 56)
	bytes[1] = byte(v6.Hi >> 48)
	bytes[2] = byte(v6.Hi >> 40)
	bytes[3] = byte(v6.Hi >> 32)
	bytes[4] = byte(v6.Hi >> 24)
	bytes[5] = byte(v6.Hi >> 16)
	bytes[6] = byte(v6.Hi >> 8)
	bytes[7] = byte(v6.Hi)
	bytes[8] = byte(v6.Lo >> 56)
	bytes[9] = byte(v6.Lo >> 48)
	bytes[10] = byte(v6.Lo >> 40)
	bytes[11] = byte(v6.Lo >> 32)
	bytes[12] = byte(v6.Lo >> 24)
	bytes[13] = byte(v6.Lo >> 16)
	bytes[14] = byte(v6.Lo >> 8)
	bytes[15] = byte(v6.Lo)
	return netip.AddrFrom16(bytes).String()
}
